File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.java
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        // Edge case: if the list has only one node, return null
        if (head == null || head.next == null) {
            return null;
        }

        // Step 1: Find the size of the linked list
        ListNode temp = head;
        int size = 0;
        while (temp != null) {
            size++;
            temp = temp.next;
        }

        // Step 2: Calculate the middle index
        int mid = size / 2;

        // Step 3: Traverse to the node just before the middle node
        temp = head;
        int t = 0;
        while (t < mid - 1) {
            temp = temp.next;
            t++;
        }

        // Step 4: Delete the middle node
        temp.next = temp.next.next;

        // Return the modified list
        return head;
    }
}

File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* deleteMiddle(struct ListNode* head) 
{
    if (!head->next) return NULL;
    struct ListNode *fast =head->next;
    struct ListNode *slow =head;

    while(fast->next!=NULL && fast->next->next!=NULL)
    {
        slow=slow->next;
        fast=fast->next->next;
    }    
    slow->next=slow->next->next;
    return head;
    
}